### PR TITLE
[sparse] Fix incorrect reverse-mode derivative for transposed COO and CSR matmat/matvec

### LINKED ISSUE #40625: jax.experimental.sparse.coo_matmat gives incorrect reverse derivative with transpose=True
### Description

`jax.experimental.sparse.coo_matmat` appears to produce an incorrect reverse-mode derivative when `transpose=True`.

In the example below, the sparse structure is fixed and valid. With `transpose=True`, the dense equivalent is `A.T @ B`. The scalar function being differentiated is equivalent to:

```
5*t + 4*(2*t - 1) + 29*(t*t + 1)
```

At `t = -24.0`, the expected first derivative is `-1379.0`. However, JAX returns `-35.0`.

### Minimal reproducible example

```
#!/usr/bin/env python3
import os
import sys
import math

os.environ.setdefault("CUDA_VISIBLE_DEVICES", "")
os.environ.setdefault("OMP_NUM_THREADS", "1")
os.environ.setdefault("JAX_PLATFORMS", "cpu")

try:
    import jax
    import jax.numpy as jnp
    import jax.experimental.sparse
except ImportError as e:
    print(f"missing dep: {e}")
    sys.exit(2)

jax.config.update("jax_enable_x64", True)

def target(t):
    data = jnp.stack((t, t * t + 1, 2 * t - 1))
    row = jnp.asarray([0, 0, 1], dtype=jnp.int32)
    col = jnp.asarray([0, 2, 1], dtype=jnp.int32)

    a = jax.experimental.sparse.COO(
        (data, row, col),
        shape=(2, 3),
        rows_sorted=True,
    )

    b = jnp.asarray([[1, 2], [3, -1]], dtype=t.dtype)
    w = jnp.asarray([[1, 2], [3, 5], [7, 11]], dtype=t.dtype)

    return jnp.sum(
        w * jax.experimental.sparse.coo_matmat(a, b, transpose=True)
    )

x = jnp.asarray(-24.0, dtype=jnp.float64)
expected = -1379.0

actual = float(jax.grad(target)(x))

print("actual:", actual)
print("expected:", expected)

if not (math.isfinite(actual) and abs(actual - expected) <= 1e-12):
    sys.exit(0)
sys.exit(1)
```

### Actual result

```
actual: -35.0
expected: -1379.0
```

### Expected result

```
-1379.0
```

The forward value is correct, but the reverse-mode derivative is not. The expected derivative follows from differentiating the dense-equivalent scalar expression:

```
5*t + 4*(2*t - 1) + 29*(t*t + 1)
```

JAX returns `-35.0` instead of `-1379.0`, suggesting that the transpose path in the sparse matrix-matrix multiplication rule is using an incorrect reverse-mode autodiff rule or is dropping part of the contribution from the transposed sparse structure.

### System info (python version, jaxlib version, accelerator, etc.)

jax: 0.10.1
jaxlib: 0.10.1
numpy: 2.4.6
python: 3.13.13
platform: Ubuntu 24.04.3 LTS, Linux x86_64
accelerator: CPU backend

Note: GPU is present, but CUDA-enabled jaxlib is not installed, so JAX falls back to CPU.

### PR DESCRIPTION
**Pull request by jdymitarai**
Fixes #40625.

### Background and Problem
In `jax.experimental.sparse`, sparse matrix-matrix and matrix-vector multiplications (`coo_matmat`, `coo_matvec`, `csr_matmat`, `csr_matvec`) support an optional `transpose: bool` flag. When `transpose=True`, the primal computes:
- Matrix-matrix: $C = A^T B$
- Matrix-vector: $y = A^T v$

For a non-zero element of $A$ at position $(row[p], col[p])$:
- In $A^T$, $col[p]$ represents the row index (matching cotangent $ct$), and $row[p]$ represents the column index (matching $B$ or $v$).
- Differentiating with respect to the non-zero sparse data of $A$ requires contracting $ct[col]$ with $B[row]$ (or $v[row]$).

Previously, the primitive transpose rules (`_coo_matmat_transpose`, `_coo_matvec_transpose`, `_csr_matmat_transpose`, `_csr_matvec_transpose`) always contracted $ct[row]$ with $B[col]$ (or $v[col]$), completely ignoring `transpose`. Consequently:
1. Whenever $row \neq col$ (e.g. non-symmetric or rectangular sparse matrices), the reverse-mode derivative with respect to the matrix data was computed with incorrect index associations, yielding wrong gradients.
2. In issue #40625, this caused `jax.grad` on a scalar loss to return `-35.0` instead of `-1379.0`.

### Solution
1. In `_coo_matvec_transpose` and `_coo_matmat_transpose` (`jax/experimental/sparse/coo.py`), swap row and col indices when `transpose=True`:
   ```python
   r, c = (col, row) if transpose else (row, col)
   ```
2. In `_csr_matvec_transpose` and `_csr_matmat_transpose` (`jax/experimental/sparse/csr.py`), swap row and col indices when `transpose=True`:
   ```python
   row, col = _csr_to_coo(indices, indptr)
   if transpose:
     row, col = col, row
   ```
3. Update `test_coo_matmul_ad` in `tests/sparse_test.py` to parameterize across `transpose=[True, False]` and add `test_csr_matmul_ad` parameterized across `transpose=[True, False]`, testing both forward-mode (`jvp`) and reverse-mode (`vjp`) autodiff with respect to both inputs (vectors/matrices and sparse non-zero data).

### Verification
- Tested locally with pytest on `tests/sparse_test.py -k "matmul_ad"`: all 20 parameterized test cases pass.
- Full COO and CSR test suite (`tests/sparse_test.py -k "coo or csr"`): all 233 passed.

### CODE REVIEW

**Comment by jdymitarai** [comment]
Closing for now to keep the PR queue focused. Thanks!
